### ISSUE TITLE
New Windows kubelet stats collection test flaking

### ISSUE THREAD

**Issue by jsturtevant** [kubernetes/kubernetes#106278]
### Which jobs are flaking?

ci-kubernetes-e2e-aks-engine-azure-master-windows-containerd

### Which tests are flaking?

Kubernetes e2e suite: [sig-windows] [Feature:Windows] Kubelet-Stats Kubelet stats collection for Windows nodes when running 3 pods should return within 10 seconds

### Since when has it been flaking?

it flaked once this week

### Testgrid link

https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-aks-engine-azure-master-windows-containerd/1457929535654727680

### Reason for failure (if possible)

```
Nov  9 05:06:31.246: Error getting kubelet stats
Unexpected error:
    <*errors.StatusError | 0xc004833360>: {
        ErrStatus: {
            TypeMeta: {Kind: "", APIVersion: ""},
            ListMeta: {
                SelfLink: "",
                ResourceVersion: "",
                Continue: "",
                RemainingItemCount: nil,
            },
            Status: "Failure",
            Message: "an error on the server (\"Internal Error: failed to list pod stats: failed to list all container stats: rpc error: code = Unknown desc = failed to convert to cri containerd stats format: failed to decode container metrics for \\\"f8b6d2a2408a6f56d0adc347729a9de5ddbadf61ea5ccba9fb06265efb962bf7\\\": windows stats is empty\") has prevented the request from succeeding (get nodes 4115k8s000:10250)",
            Reason: "InternalError",
            Details: {
                Name: "4115k8s000:10250",
                Group: "",
                Kind: "nodes",
                UID: "",
                Causes: [
                    {
                        Type: "UnexpectedServerResponse",
                        Message: "Internal Error: failed to list pod stats: failed to list all container stats: rpc error: code = Unknown desc = failed to convert to cri containerd stats format: failed to decode container metrics for \"f8b6d2a2408a6f56d0adc347729a9de5ddbadf61ea5ccba9fb06265efb962bf7\": windows stats is empty",
                        Field: "",
                    },
                ],
                RetryAfterSeconds: 0,
            },
            Code: 500,
        },
    }
```

### Anything else we need to know?

https://github.com/kubernetes/kubernetes/pull/105854 was merged yesterday.

fyi @NikhilSharmaWe

### Relevant SIG(s)

/sig windows

**Comment by NikhilSharmaWe** [comment]
@jsturtevant How can I help in this issue.

**Comment by jsturtevant** [comment]
See if you can reproduce it and identify why it is flaking. It has flaked a couple more times since the initial report.  Tagged you since you were the author, I'll be able to take a look next week if not.  If you need help reach out in slack.

**Comment by jsturtevant** [comment]
had another flake here: https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-aks-engine-azure-master-windows-containerd/1465027326419931136

**Comment by claudiubelu** [comment]
I've investigated this a bit, and found the source of the issue, and I could even reproduce it.

So, this failure will mostly occur on Windows test jobs in which we run tests in parallel. The failure is not caused by the test itself, but rather by other tests running in parallel.

What's happening: the test will get the node stats a couple of times. That will also include the pod / container stats [1], which will ultimately call ``p.runtimeService.ListContainerStats(&runtimeapi.ContainerStatsFilter{})`` [2], which means that it'll get all the stats for all the containers from containerd.

However, some errors are encountered while getting these stats in containerd, and more specifically, getting the stats from containers that are already dead / stopped. This can be observed in the test, kubelet, and containerd logs for the job [3]. We can see in the test logs this failure:

```
Message: "an error on the server (\"Internal Error: failed to list pod stats: failed to list all container stats: rpc error: code = Unknown desc = failed to convert to cri containerd stats format: failed to decode container metrics for \\\"67d8d77087a2369383a2eb4c8189428f1eae6607121c3f98aa9176e4c87c4958\\\": windows stats is empty\") has prevented the request from succeeding (get nodes 3822k8s000:10250)",
```

Looking at the containerd logs, we can see that that container ID corresponds to an ``agnhost-container`` container (the containers spawned by this test has the names ``stat-container``), which indicates that it belongs to another test.

```
time="2022-01-09T21:02:37.265952200Z" level=info msg="CreateContainer within sandbox \"71851f88215ad40fc7739cb669057e78e2605bd6984b6ffd070208a03c27f528\" for &ContainerMetadata{Name:agnhost-container,Attempt:0,} returns container id \"67d8d77087a2369383a2eb4c8189428f1eae6607121c3f98aa9176e4c87c4958\""
time="2022-01-09T21:02:37.284960400Z" level=info msg="StartContainer for \"67d8d77087a2369383a2eb4c8189428f1eae6607121c3f98aa9176e4c87c4958\""
```

Later on, we can see that this container is stopped, and immediately afterwards we see the error above:

```
time="2022-01-09T21:02:57.031455700Z" level=info msg="StopContainer for \"67d8d77087a2369383a2eb4c8189428f1eae6607121c3f98aa9176e4c87c4958\" with timeout 1 (s)"
time="2022-01-09T21:02:57.066062900Z" level=info msg="Stop container \"67d8d77087a2369383a2eb4c8189428f1eae6607121c3f98aa9176e4c87c4958\" with signal terminated"
time="2022-01-09T21:02:58.509614800Z" level=info msg="Kill container \"67d8d77087a2369383a2eb4c8189428f1eae6607121c3f98aa9176e4c87c4958\""
time="2022-01-09T21:02:59.021444900Z" level=error msg="ListContainerStats failed" error="failed to convert to cri containerd stats format: failed to decode container metrics for \"67d8d77087a2369383a2eb4c8189428f1eae6607121c3f98aa9176e4c87c4958\": windows stats is empty"
time="2022-01-09T21:02:59.456415500Z" level=error msg="ListContainerStats failed" error="failed to convert to cri containerd stats format: failed to decode container metrics for \"67d8d77087a2369383a2eb4c8189428f1eae6607121c3f98aa9176e4c87c4958\": windows stats is empty"
```

The error case described above can be seen over and over through most / all test runs in the containerd logs.

This issue can also be reproduced after running containerd ``cri-integration`` tests a few times [4], with these changes:

```
diff --git a/integration/container_stats_test.go b/integration/container_stats_test.go
index 2f49d1b0d..17e6505ee 100644
--- a/integration/container_stats_test.go
+++ b/integration/container_stats_test.go
@@ -49,7 +49,7 @@ func TestContainerStats(t *testing.T) {
        }()
        require.NoError(t, runtimeService.StartContainer(cn))
        defer func() {
-               assert.NoError(t, runtimeService.StopContainer(cn, 10))
+               runtimeService.StopContainer(cn, 10)
        }()

        t.Logf("Fetch stats for container")
@@ -67,6 +67,17 @@ func TestContainerStats(t *testing.T) {

        t.Logf("Verify stats received for container %q", cn)
        testStats(t, s, containerConfig)
+
+       t.Logf("Verify that there are no errors when collecting stats from a stopped container.")
+       go func() {
+               assert.NoError(t, runtimeService.StopContainer(cn, 10))
+       }()
+
+       for i := 0; i < 50; i++ {
+               _, err = runtimeService.ListContainerStats(&runtime.ContainerStatsFilter{})
+               assert.NoError(t, err)
+               time.Sleep(100 * time.Millisecond)
+       }
 }

 // Test to verify if the consumed stats are correct.
```

This issue should be solved on the containerd side, since we can't really do much about it on the kubelet side (we'd have to query every ``Running`` container ourselves, which isn't feasible). IMO, when querying **all** the container stats, it should't end up in a general error if it failed to query **one** container (but it would make sense if only that one container is querried specifically). On the containerd side, we shouldn't query stopped / not running containers since they don't consume stats in the first place (or just return zeroes instead).

[1] https://github.com/kubernetes/kubernetes/blob/f86ee5649b2e127a6471bbc896614ecd32ff57c7/pkg/kubelet/server/stats/summary.go#L134
[2] https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/cri_stats_provider.go#L307
[3] https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-aks-engine-azure-master-windows-containerd/1480270936501391360
[4] https://paste.ubuntu.com/p/TrVHC3F5mS/

**Comment by jsturtevant** [comment]
Do we have an open issue in containerd we can cross link here?

**Comment by gabriel-samfira** [comment]
> Do we have an open issue in containerd we can cross link here?

I opened one here:

https://github.com/containerd/containerd/issues/6436

**Comment by dcantah** [comment]
Thanks @gabriel-samfira for the fix! Once [this](https://github.com/microsoft/hcsshim/pull/1277) is in I'll need to cut a new 0.9.x tag in hcsshim and get it into containerd. Optimistically the fix will be present in Containerd 1.6

**Comment by marosset** [comment]
/close
We recently updated some test configs for windows to run 1.6.0-rc.1.
Ex - https://testgrid.k8s.io/sig-windows-master-release#aks-engine-windows-containerd-master
